[Bug] User.assignSupervisor will return false

## Prerequisites & Information

- **Area Path:** usermanagementsystem\.datastructure\.User
- **Test Case Class:** usermanagementsystem\.test\TestUser
- **Release Version:** V1.0
- **Severity:** HIGH
- **Priority:** HIGH 
- **Reported On:** 28 Nov 2017
- **Reason:** Defect
- **Status:** Waiting for fix
- **Environment:** jdk1.8.0_152, JUnit4, Intellij 2017.2.5

## Description:
The actual result is different with the expected result. The expected result should be the correct one. After the test case run, the result should be false but the actual result is true.

## Steps To Reproduce:

1. Run the test case - TestUser
2. Run the testUserAssignSupervisor_2 Method
3. The result will be shown in console
4. The error message will be displayed.


Expected result: False
Actual Result: True

exclude patterns:
java.lang.AssertionError: 
Expected :false
Actual   :true

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/controller/AdminController.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/controller/AdminController.java
@@ -29,7 +29,9 @@ public class AdminController extends UserController {
     /**
      * Singleton AdminController
      *
-     * @param admin admin User
+     * @param admin       admin User
+     * @param users       all User
+     * @param supervisors all Supervisor
      * @return instance AdminController
      * @throws ExControllerInitWithNull the input param cannot be null
      */

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/Supervisor.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/Supervisor.java
@@ -3,6 +3,7 @@ package usermanagementsystem.datastructure;
 import java.util.Hashtable;
 
 import usermanagementsystem.datastructure_interface.*;
+import usermanagementsystem.exception.ExIsNullOrEmpty;
 
 /**
  * The data type Supervisor. It can be parsed by Gson from json
@@ -84,9 +85,11 @@ public class Supervisor extends User implements ISupervisorInfo {
          * The build method to build new Supervisor object
          *
          * @return Supervisor object
+         * @throws ExIsNullOrEmpty important data field cannot be null or empty
          */
         @Override
-        public Supervisor build() {
+        public Supervisor build() throws ExIsNullOrEmpty {
+            validation();
             return new Supervisor(userName, password, gender, position, staffId, email, departmentOf, supervisor, isAdmin);
         }
     }

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/Supervisor.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/Supervisor.java
@@ -80,6 +80,11 @@ public class Supervisor extends User implements ISupervisorInfo {
      * The object builder helps to create new Supervisor object. Inherits <b>UserBuilder</b>
      */
     public static class SupervisorBuilder extends UserBuilder {
+        /**
+         * The build method to build new Supervisor object
+         *
+         * @return Supervisor object
+         */
         @Override
         public Supervisor build() {
             return new Supervisor(userName, password, gender, position, staffId, email, departmentOf, supervisor, isAdmin);

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
@@ -351,7 +351,14 @@ public class User implements IUserInfo, Comparable<User> {
          *
          * @return User object
          */
-        public User build() {
+        public User build() throws ExIsNullOrEmpty {
+            //verify
+            strIsNullOrEmpty("userName", userName);
+            strIsNullOrEmpty("password", password);
+            strIsNullOrEmpty("email", email);
+            if (gender == null) throw new ExIsNullOrEmpty("gender");
+            if (position == null) throw new ExIsNullOrEmpty("position");
+            if (departmentOf == null) throw new ExIsNullOrEmpty("departmentOf");
             return new User(userName, password, gender, position, Calendar.getInstance().getTimeInMillis(), email, departmentOf, supervisor, isAdmin);
         }
     }

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
@@ -184,7 +184,7 @@ public class User implements IUserInfo, Comparable<User> {
      * @return boolean
      */
     public boolean assignSupervisor(ISupervisorInfo supervisor) {
-        if (this.supervisor == null) {
+        if (this.supervisor == null && supervisor != null) {
             this.supervisor = supervisor;
             return true;
         }

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
@@ -247,7 +247,7 @@ public class User implements IUserInfo, Comparable<User> {
         EnumGender gender;
         EnumPosition position;
         EnumDepartment departmentOf;
-        boolean isAdmin;
+        boolean isAdmin = false;
         ISupervisorInfo supervisor = null;
 
         /**
@@ -346,13 +346,10 @@ public class User implements IUserInfo, Comparable<User> {
             return this;
         }
 
-//        public UserBuilder supervisor(ISupervisorInfo supervisor) {
-//            this.supervisor = supervisor;
-//            return this;
-//        }
-
         /**
          * The build method to build new User object
+         *
+         * @return User object
          */
         public User build() {
             return new User(userName, password, gender, position, Calendar.getInstance().getTimeInMillis(), email, departmentOf, supervisor, isAdmin);

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
@@ -242,7 +242,7 @@ public class User implements IUserInfo, Comparable<User> {
     public static class UserBuilder {
         String userName;
         String password;
-        long staffId;
+        long staffId = Calendar.getInstance().getTimeInMillis();
         String email;
         EnumGender gender;
         EnumPosition position;
@@ -359,7 +359,7 @@ public class User implements IUserInfo, Comparable<User> {
             if (gender == null) throw new ExIsNullOrEmpty("gender");
             if (position == null) throw new ExIsNullOrEmpty("position");
             if (departmentOf == null) throw new ExIsNullOrEmpty("departmentOf");
-            return new User(userName, password, gender, position, Calendar.getInstance().getTimeInMillis(), email, departmentOf, supervisor, isAdmin);
+            return new User(userName, password, gender, position, staffId, email, departmentOf, supervisor, isAdmin);
         }
     }
 }

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/datastructure/User.java
@@ -347,18 +347,28 @@ public class User implements IUserInfo, Comparable<User> {
         }
 
         /**
-         * The build method to build new User object
+         * Validate important data field
          *
-         * @return User object
+         * @throws ExIsNullOrEmpty important data field cannot be null or empty
          */
-        public User build() throws ExIsNullOrEmpty {
-            //verify
+        protected void validation() throws ExIsNullOrEmpty {
             strIsNullOrEmpty("userName", userName);
             strIsNullOrEmpty("password", password);
             strIsNullOrEmpty("email", email);
             if (gender == null) throw new ExIsNullOrEmpty("gender");
             if (position == null) throw new ExIsNullOrEmpty("position");
             if (departmentOf == null) throw new ExIsNullOrEmpty("departmentOf");
+        }
+
+        /**
+         * The build method to build new User object
+         *
+         * @return User object
+         * @throws ExIsNullOrEmpty important data field cannot be null or empty
+         */
+        public User build() throws ExIsNullOrEmpty {
+            //verify
+            validation();
             return new User(userName, password, gender, position, staffId, email, departmentOf, supervisor, isAdmin);
         }
     }

--- a/cs3343.group18.usermanagementsystem/src/usermanagementsystem/test/TestUser.java
+++ b/cs3343.group18.usermanagementsystem/src/usermanagementsystem/test/TestUser.java
@@ -10,7 +10,6 @@ import static org.junit.Assert.assertEquals;
 
 public class TestUser {
 
-
     @Test
     public void testUserGetUserName() throws ExInvaildEnumValue, ExIsNullOrEmpty {
         User userTest;
@@ -57,7 +56,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserAddPermission() {
+    public void testUserAddPermission() throws ExIsNullOrEmpty {
         User userTest;
         User.UserBuilder builder = new User.UserBuilder();
         userTest = builder.build();
@@ -68,7 +67,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserRemovePermission() {
+    public void testUserRemovePermission() throws ExIsNullOrEmpty {
         User userTest;
         User.UserBuilder builder = new User.UserBuilder();
         userTest = builder.build();
@@ -80,7 +79,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserShowAllPermission() {
+    public void testUserShowAllPermission() throws ExIsNullOrEmpty {
         User userTest;
         String expected;
         User.UserBuilder builder = new User.UserBuilder();
@@ -92,7 +91,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserEamil() throws ExIsNullOrEmpty {
+    public void testUserEmail() throws ExIsNullOrEmpty {
         User userTest;
         String expected;
         User.UserBuilder builder = new User.UserBuilder();
@@ -104,7 +103,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserPosition() {
+    public void testUserPosition() throws ExIsNullOrEmpty {
         User userTest;
         EnumPosition expected;
         User.UserBuilder builder = new User.UserBuilder();
@@ -115,7 +114,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserDepartment() {
+    public void testUserDepartment() throws ExIsNullOrEmpty {
         User userTest;
         EnumDepartment expected;
         User.UserBuilder builder = new User.UserBuilder();
@@ -126,7 +125,7 @@ public class TestUser {
     }
 
     @Test
-    public void testUserIsAdmin() {
+    public void testUserIsAdmin() throws ExIsNullOrEmpty {
         User userTest;
         boolean expected;
         User.UserBuilder builder = new User.UserBuilder();
@@ -171,11 +170,11 @@ public class TestUser {
         Supervisor expected;
         User.UserBuilder builder = new User.UserBuilder();
         builder.userName("supervisor")
-        .departmentOf(EnumDepartment.Technology)
-        .password("abc")
-        .position(EnumPosition.ProductOwner)
-        .email("supervisor@test.com")
-        .isAdmin(true);
+                .departmentOf(EnumDepartment.Technology)
+                .password("abc")
+                .position(EnumPosition.ProductOwner)
+                .email("supervisor@test.com")
+                .isAdmin(true);
         userTest = builder.build();
         expected = userTest.toSupervisor();
         assertEquals(true, expected.isSupervisor());
@@ -196,6 +195,7 @@ public class TestUser {
         expected = userTest.assignSupervisor(userTest.toSupervisor());
         assertEquals(true, expected);
     }
+
     @Test
     public void testUserAssignSupervisor_2() throws ExIsNullOrEmpty {
         User userTest;

